dev_apps/uPnP-test/js/app.js

```diff
@@ -0,0 +1,277 @@
+  var audioPlayer;
+  var videoPlayer;
+  var imagePlayer;
+  var discoverButton;
+  var currentPlayer;
+  var playerToggler;
+  var folderList;
+
+  function debugLog(msg, level) {
+    if (!DEBUG && level == 'debug') {
+      return;
+    }
+    var logEl = document.createElement('div');
+    logEl.textContent = msg;
+    debugEl.appendChild(logEl);
+  }
+
+  function toggleFolder(evt) {
+    evt.preventDefault();
+    var elem = evt.target;
+
+    if (elem.nextSibling && elem.nextSibling.tagName == 'DIV') {
+      // folder is opened. Close it.
+      elem.parentElement.removeChild(elem.nextSibling);
+      elem.classList.remove('opened');
+    } else {
+      // folder is closed. Open it.
+      debugLog(elem.serviceId, 'debug');
+      elem.classList.add('opened');
+      browseFolder(
+        elem.dataset.serviceId, elem.hash.substr(1), evt.target);
+    }
+  }
+
+  function togglePlayer(value) {
+    if (typeof value != 'boolean') {
+      value = currentPlayer.classList.contains('hide');
+    }
+    audioPlayer.classList.add('hide');
+    videoPlayer.classList.add('hide');
+    imagePlayer.classList.add('hide');
+    if (value) {
+      currentPlayer.classList.remove('hide');
+    }
+  }
+
+  function switchPlayer(type) {
+    if (type == 'audio') {
+      currentPlayer = audioPlayer;
+    } else if (type == 'video') {
+      currentPlayer = videoPlayer;
+    } else if (type == 'image') {
+      currentPlayer = imagePlayer;
+    }
+    togglePlayer(true);
+  }
+
+  function playFile(evt) {
+    evt.preventDefault();
+    var fileType = evt.target.dataset.mime.split('/')[0].toLowerCase();
+    var fileFormat = evt.target.dataset.mime.split('/')[1].toLowerCase();
+//    if (fileFormat == 'mpeg' || fileFormat == 'mp4') {
+//      window.open(evt.target.href, 'mpegPopup');
+//      return;
+//    }
+
+    if (evt.target.href.substr(-3) == 'ogv') {
+      fileFormat = 'video/ogg';
+    } else if (evt.target.href.substr(-3) == 'ogg') {
+      fileFormat = 'audio/ogg';
+    }
+    switchPlayer(fileType);
+
+    currentPlayer.hidden = false;
+    currentPlayer.src = evt.target.href;
+    currentPlayer.setAttribute('type', fileFormat);
+    currentPlayer.oncanplay = function() {
+      this.play();
+    };
+  }
+
+  function browseFolder(serviceId, folderId, folderElement) {
+    var mediaServer = savedServices[serviceId];
+    if (!mediaServer) {
+      return;
+    }
+
+    folderId = folderId || '';
+
+    mediaServer.browse(folderId).then(function(response) {
+      if (!response || !response.data) {
+        return;
+      }
+
+
+      var data = response.data.Result;
+      debugLog(data);
+
+      if (data.indexOf('xmlns:dlna') == -1) {
+        data = data.replace('<DIDL-Lite ',
+          '<DIDL-Lite xmlns:dlna="urn:schemas-dlna-org:device-1-0" ');
+      }
+      var parser = new DOMParser();
+      var serializer = new XMLSerializer();
+      var xmlResponse = parser.parseFromString(response.data.Result, 'application/xml');
```

Comment:
XML response containing one media file is getting rendered. Seeing XML not well formed error if response XML is big and contains more than one media files. Looking into it.